FeatureFlags: define features outside settings.Cfg (take 3)

ryantxu opened the pull request:
This is new version of: #43692 and #44404

This version is less ambitious and keeps the parallel `licensedFeatures` registry untouched.   

This will let us punt on features that also have their enabled state defined in settings.Cfg

ryantxu committed:
Revert "FeatureFlags: Revert managing feature flags outside of settings.Cfg (#44382)"

This reverts commit 65bdb3a899848cb92b8d6d2ca263a5ed3f9cd873.

ryantxu committed:
list features

ryantxu committed:
list features

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags-take-2

ryantxu committed:
update licensing api

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags-take-2

ryantxu committed:
remove license integration

DanCech commented on pkg/api/http_server.go:
```diff
@@ -76,6 +77,7 @@ type HTTPServer struct {
 	Bus                       bus.Bus
 	RenderService             rendering.Service
 	Cfg                       *setting.Cfg
+	Features                  *featuremgmt.FeatureManager
```
Do we need FeatureManager here or could we use FeatureToggles?

ryantxu replied:
there is an [HTTP endpoint ](https://github.com/grafana/grafana/pull/44443/files#diff-67a6e36014be865053c43d38b2091bd2d3f17a91b117d97c0f5244cd7d71127dR464) that can list details and should eventually support changing state (ie, writing to DB or temporary change etc)

Perhaps FeatureManager could extend FeatureToggles so we do not need: `hs.Features.Toggles().IsLivePipelineEnabled()`


If we drop the `IsXXXEnabled()` helpers, this would be much easier, but the checks become a bit more verbose (maybe better/ok?):
```
s.features.IsEnabled(featuremgmt.FlagLivePipeline)
```

DanCech replied:
yeah dropping the helper functions might make sense, we're already importing featuremgmt for the struct defs so that would make things cleaner and also allow us to accept a FeatureToggles interface that could be fulfilled by FeatureMgmt, and we'd still have static checking to make sure flags are valid.

ryantxu replied:
@DanCech just pushed a big refactor that uses a constant for everything

DanCech commented on pkg/services/sqlstore/org_users_test.go:
```diff
@@ -61,7 +62,7 @@ func TestSQLStore_GetOrgUsers(t *testing.T) {
 
 	store := InitTestDB(t)
-	store.Cfg.FeatureToggles = map[string]bool{"accesscontrol": true}
+	store.Cfg.IsFeatureToggleEnabled = featuremgmt.WithToggles("accesscontrol").IsEnabled
```
we repeat `"accesscontrol"` a fair bit, should we maybe add that as a FLAG value?

ryantxu replied:
done

ryantxu committed:
use constants for flags

DanCech commented on pkg/services/featuremgmt/toggles_gen_test.go:
```diff
@@ -100,10 +99,8 @@ func asCamelCase(key string) string {
-func (ft *FeatureToggles) Is{{.CamleCase}}Enabled() bool {
-	return ft.manager.IsEnabled("{{.Flag.Name}}")
-}
+{{"\t"}}// Flag{{.CamleCase}} {{.Flag.Name}}{{.Ext}}
```
Was CamleCase a typo or intentional?
```suggestion
{{"\t"}}// Flag{{.CamelCase}} {{.Flag.Name}}{{.Ext}}
```

ryantxu replied:
you are too kind :facepalm:

DanCech commented:
This is feeling really nice now!  It also seems like we can now pass featuremgmt.FlagX values in the tests that are still using strings

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags-take-3

ryantxu committed:
use constants

ryantxu committed:
use constants

ryantxu committed:
add validatedQueries

ryantxu committed:
add validatedQueries

ryantxu committed:
use features in cli

ryantxu committed:
typos and cleanup

leeoniya committed:
Formatters: speed up toDuration() (#44345)

Clarity-89 committed:
NavModel: Enable adding suffix elements to tabs (#44155)

* Trigger extra events

* Extend html attributes

* Add suffix to tabs

* Add upgrade routes

* suffix => highlightText

* suffix => suffixText

* Add size prop

* Update prop name

* Convert tabSuffix to a component

marefr committed:
Provisioning: Fix duplicate validation when multiple organizations have been configured (#44151)

Fix duplicate validation when multiple organizations have been configured. 
This makes sure that all duplicate validation is being made for each organization.

Fixes #44126

mtanda committed:
Cloud Monitoring: Reduce request size when listing labels (#44365)

marefr committed:
Chore: Implements log15 terminal logger (#44447)

Nicer development/terminal output based on log15 terminal logger.

mortenaa committed:
MSSQL: Configuration of certificate verification for TLS connection (#31865)

Fixes #24589

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

AgnesToulet commented on pkg/services/featuremgmt/registry.go:
```diff
@@ -0,0 +1,125 @@
+var (
+	// Register each toggle here
+	standardFeatureFlags = []FeatureFlag{
+		{
+			Name:            "caching",
+			Description:     "Temporarily store data source query results.",
+			State:           FeatureStateStable,
+			DocsURL:         "https://grafana.com/docs/grafana/latest/enterprise/query-caching/",
+			RequiresLicense: true,
+		},
```
```suggestion
		{
			Name:            "caching",
			Description:     "Temporarily store data source query results.",
			State:           FeatureStateStable,
			DocsURL:         "https://grafana.com/docs/grafana/latest/enterprise/query-caching/",
			RequiresLicense: true,
		},
```
This feature toggle has been removed in https://github.com/grafana/grafana-enterprise/pull/1270

AgnesToulet commented on pkg/services/featuremgmt/manager_test.go:
```diff
@@ -0,0 +1,77 @@
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
```
```suggestion
	"github.com/stretchr/testify/assert"
```
Nit but I think this file should only use `assert` for the checks.

AgnesToulet commented on pkg/services/serviceaccounts/manager/service_test.go:
```diff
@@ -5,31 +5,29 @@ import (
-			cfg:   cfg,
-			store: svcMock,
-			log:   log.New("serviceaccounts-manager-test"),
+			features: featuremgmt.WithFeatures("service-accounts", false),
```
```suggestion
			features: featuremgmt.WithFeatures(featuremgmt.FlagServiceAccounts, false),
```

AgnesToulet commented on pkg/services/featuremgmt/service_test.go:
```diff
@@ -0,0 +1,84 @@
+	require.NotNil(t, mgmt)
+
+	// Enterprise features do not fall though automatically
+	require.False(t, mgmt.IsEnabled("a.yes.default"))
+	require.False(t, mgmt.IsEnabled("a.yes")) // licensed, but not enabled
```
```suggestion
	assert.False(t, mgmt.IsEnabled("a.yes.default"))
	assert.False(t, mgmt.IsEnabled("a.yes")) // licensed, but not enabled
```
Nit

AgnesToulet reviewed (commented):
Looks pretty good! I'll test it after reviewing the Enterprise one. Left a few mostly minor comments.

renovate[bot] committed:
Update dependency eslint-plugin-jsdoc to v37.7.0 (#44459)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency @visx/tooltip to v2.8.0 (#44458)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

gamab committed:
AccessControl: Implement teams resource service (#43951)

* AccessControl: cover team permissions

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Add background service as a consumer to resource_services

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Define actions in roles.go

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Remove action from accesscontrol model

 Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* As suggested by kalle

* move some changes from branch to the skeleton PR

* Add background service as a consumer to resource_services

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* moving resourceservice to the main wire file pt2

* move team related actions so that they can be reused

* PR feedback

* fix

* typo

* Access Control: adding hooks for team member endpoints (#43991)

* AccessControl: cover team permissions

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Add background service as a consumer to resource_services

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Define actions in roles.go

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Remove action from accesscontrol model

 Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* As suggested by kalle

* add access control to list and add team member endpoint, and hooks for adding team members

* member permission type is 0

* add ID scope for team permission checks

* add more team actions, use Member for member permission name

* protect team member update endpoint with FGAC permissions

* update SQL functions for teams and the corresponding tests

* also protect team member removal endpoint with FGAC permissions and add a hook to permission service

* a few small fixes, provide team permission service to test setup

* AccessControl: cover team permissions

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Add background service as a consumer to resource_services

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Define actions in roles.go

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* Remove action from accesscontrol model

 Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

* As suggested by kalle

* move some changes from branch to the skeleton PR

* remove resource services from wireexts

* remove unneeded actions

* linting fix

* remove comments

* feedback fixes

* feedback

* simplifying

* remove team member within the same transaction

* fix a mistake with the error

* call the correct sql fction

* linting

* Access control: tests for team member endpoints (#44177)

* tests for team member endpoints

* clean up and fix the tests

* fixing tests take 2

* don't import enterprise test license

* don't import enterprise test license

* remove unused variable

Co-authored-by: gamab <gabi.mabs@gmail.com>
Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

Co-authored-by: ievaVasiljeva <ieva.vasiljeva@grafana.com>

gabor committed:
CODEOWNERS: update based on o11y squad changes (#44474)

* CODEOWNERS: update based on o11y squad changes

* add back by-mistake deleted part

* move logs-panel to logs-traces-squad

zserge committed:
Chore: Remove bus from ngalert (#44465)

* pass notification service down to the notifiers

* add ns to all notifiers

* remove bus from ngalert notifiers

* use smaller interfaces for notificationservice

* attempt to fix the tests

* remove unused struct field

* simplify notification service mock

* trying to resolve issues in the tests

* make linter happy

* make linter even happier

* linter, you are annoying

ryantxu committed:
Update pkg/services/secrets/manager/helpers.go

Co-authored-by: Agnès Toulet <35176601+AgnesToulet@users.noreply.github.com>

ryantxu committed:
Update pkg/services/secrets/manager/helpers.go

Co-authored-by: Agnès Toulet <35176601+AgnesToulet@users.noreply.github.com>

ryantxu committed:
Update pkg/services/sqlstore/migrations/migrations.go

Co-authored-by: Agnès Toulet <35176601+AgnesToulet@users.noreply.github.com>

ryantxu committed:
merge main

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44443) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44443) before we can accept your contribution.<br/>**9** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: ryantxu<br/>:white_check_mark: Clarity-89<br/>:white_check_mark: leeoniya<br/>:white_check_mark: mtanda<br/>:white_check_mark: mortenaa<br/>:white_check_mark: marefr<br/>:white_check_mark: gamab<br/>:white_check_mark: zserge<br/>:white_check_mark: gabor<br/>:x: renovate[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44443) it.</sub>

ryantxu committed:
Update pkg/services/sqlstore/org_users.go

Co-authored-by: Agnès Toulet <35176601+AgnesToulet@users.noreply.github.com>

ryantxu committed:
Update pkg/services/serviceaccounts/api/api_test.go

Co-authored-by: Agnès Toulet <35176601+AgnesToulet@users.noreply.github.com>

ryantxu commented on pkg/setting/setting_unified_alerting.go:
```diff
@@ -79,7 +79,7 @@ func (cfg *Cfg) readUnifiedAlertingEnabledSetting(section *ini.Section) (*bool,
 	if err != nil {
 		// TODO: Remove in Grafana v9
-		if cfg.FeatureToggles["ngalert"] {
+		if cfg.IsFeatureToggleEnabled("ngalert") {
```
I'm not sure on this one -- and will punt until after this is merged.  The usage is very small and we don't want to advertise it anywhere.

ryantxu committed:
remove caching flag

DanCech reviewed (approved):
LGTM
